recipes/lora_finetune_distributed.py

```diff
@@ -307,16 +303,40 @@ def _setup_model(
                 base_model_state_dict_keys=base_model_state_dict.keys(),
             )
 
-            # Load both the base model weights and (if available) the adapter weights. Both
-            # of this should happen only on Rank 0
-            model.load_state_dict(base_model_state_dict, strict=False)
-            if lora_weights_state_dict:
-                model.load_state_dict(lora_weights_state_dict, strict=False)
+        self.adapter_params = get_adapter_params(model)
+        set_trainable_params(model, self.adapter_params)
 
-        else:
-            # For non-zero ranks, load the model on meta device
-            with utils.set_default_dtype(self._dtype), torch.device("meta"):
-                model = config.instantiate(cfg_model)
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                model, auto_wrap_policy={modules.TransformerDecoderLayer}
+            )
+
+        for m in reversed(list(model.modules())):
```

Comment:
My noob understanding here is that you want to account for the Linear module before the TransformerDecoderLayer - is that right? If so, is there a better way to do this? It wasn't immediately obvious to me that reversing this list gets me that outcome - but maybe that's because I haven't played around with the modules function? Disregard if this is a pretty standard way to achieve this

Reply:
I wanted to quickly mention my opinion:

What we want is a post-order traversal of the modules so that we visit lower modules before higher ones. `nn.Module.modules()` always gives _reverse_ post-order, so reversing it gives post-order.

FSDP1 hides all of this under the `auto_wrap_policy` argument, which does the post-order traversal for the user. Personally, I did not want to have an `auto_wrap_policy` argument for FSDP2 because the name is a misnomer. Auto wrapping is not automatically choosing which modules to wrap for the user -- rather, it just performing a post-order traversal and applying a user-defined policy to determine if a module is wrapped or not. What I am open to though is having some other higher-level utility (say, `apply_fsdp()`) that does the same thing as `auto_wrap_policy` but exists outside of the function `fully_shard`.